services/typeddata/hash.go

```diff
@@ -0,0 +1,214 @@
+package typeddata
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/big"
+	"reflect"
+	"sort"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+var (
+	bytes32Type, _ = abi.NewType("bytes32")
+)
+
+func deps(target string, types Types) []string {
+	unique := map[string]struct{}{}
+	unique[target] = struct{}{}
+	visited := []string{target}
+	deps := []string{}
+	for len(visited) > 0 {
+		current := visited[0]
+		fields := types[current]
+		for i := range fields {
+			f := fields[i]
+			if _, defined := types[f.Type]; defined {
+				if _, exist := unique[f.Type]; !exist {
+					visited = append(visited, f.Type)
+					unique[f.Type] = struct{}{}
+				}
+			}
+		}
+		visited = visited[1:]
+		deps = append(deps, current)
+	}
+	sort.Slice(deps[1:], func(i, j int) bool {
+		return deps[1:][i] < deps[1:][j]
+	})
+	return deps
+}
+
+func typeString(target string, types Types) string {
+	b := new(bytes.Buffer)
+	for _, dep := range deps(target, types) {
+		b.WriteString(dep)
+		b.WriteString("(")
+		fields := types[dep]
+		first := true
+		for i := range fields {
+			if !first {
+				b.WriteString(",")
+			} else {
+				first = false
+			}
+			f := fields[i]
+			b.WriteString(f.Type)
+			b.WriteString(" ")
+			b.WriteString(f.Name)
+		}
+		b.WriteString(")")
+	}
+	return b.String()
+}
+
+func typeHash(target string, types Types) (rst common.Hash) {
+	return crypto.Keccak256Hash([]byte(typeString(target, types)))
+}
+
+func encodeData(target string, data map[string]interface{}, types Types) (rst common.Hash, err error) {
+	fields := types[target]
+	typeh := typeHash(target, types)
+	args := abi.Arguments{{Type: bytes32Type}}
+	vals := []interface{}{typeh}
+	for i := range fields {
+		f := fields[i]
+		val, typ, err := toABITypeAndValue(f, data, types)
+		if err != nil {
+			return rst, err
+		}
+		vals = append(vals, val)
+		args = append(args, abi.Argument{Name: f.Name, Type: typ})
+	}
+	packed, err := args.Pack(vals...)
+	if err != nil {
+		return rst, err
+	}
+	return crypto.Keccak256Hash(packed), nil
+}
+
+func toABITypeAndValue(f Field, data map[string]interface{}, types Types) (val interface{}, typ abi.Type, err error) {
```

Comment:
This function is super important, but already big and complex enough. Also, EIP suggests that atomic types can be added in the future, which will lead to even more complication.

Maybe it can be broken down into smaller and more digestible pieces. Maybe even deserve its own file (`abi.go`)?

Reply:
but thats about **new** atomic types, which is true ofcourse, but tbh to me this function seems really simple. it is just several conditions that cast values from json to types expected by abi.
can you give an example what you thinking about and how it will help?